File: languages/c/gbdt-par.c
// gbdt-par: parallel (wall-clock scaling track) gbdt inference. Same result as
// serial gbdt.c for ANY core count. Invocation: gbdt-par <cores> <n>.
// Sample-band decomposition: the ensemble (read-only) is built serially with the
// pinned LCG; each worker traverses all B trees for its sample band and writes its
// own acc[i] (disjoint). The poly-hash + total are then computed in a single SERIAL
// pass over acc[0..N-1] (same sample order as serial) -> bit-identical checksum.
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

static long long now_ns(void) { struct timespec t; clock_gettime(CLOCK_MONOTONIC, &t); return (long long)t.tv_sec * 1000000000LL + t.tv_nsec; }

#define P          1000000007L
#define D          8
#define B          200
#define F          8
#define NODES      511
#define LEAF_START 255

static long lcg(long s) { return (s * 1103515245L + 12345L) & 0x7fffffffL; }

static int N;
static int *feat, *thr, *leafval, *sample;
static long *accArr;

typedef struct { int lo, hi; } Band;

static void *worker(void *arg) {
    Band *b = (Band *)arg;
    for (int i = b->lo; i < b->hi; i++) {
        int sbase = i * F;
        long acc = 0;
        for (int bb = 0; bb < B; bb++) {
            int tbase = bb * NODES;
            int node = 0;
            for (int d = 0; d < D; d++) {
                if (sample[sbase + feat[tbase + node]] <= thr[tbase + node])
                    node = 2 * node + 1;
                else
                    node = 2 * node + 2;
            }
            acc += leafval[tbase + node];
        }
        accArr[i] = acc;
    }
    return NULL;
}

int main(int argc, char **argv) {
    int cores = argc > 1 ? atoi(argv[1]) : 1;
    N = argc > 2 ? atoi(argv[2]) : 5000;
    if (cores < 1) cores = 1;
    if (cores > N) cores = N;
    if (cores > 256) cores = 256;

    feat    = malloc((size_t)B * NODES * sizeof(int));
    thr     = malloc((size_t)B * NODES * sizeof(int));
    leafval = malloc((size_t)B * NODES * sizeof(int));
    sample  = malloc((size_t)N * F * sizeof(int));
    accArr  = malloc((size_t)N * sizeof(long));

    long s = 42;
    for (int b = 0; b < B; b++) {
        int base = b * NODES;
        for (int node = 0; node < LEAF_START; node++) {
            s = lcg(s); feat[base + node] = (int)(s % F);
            s = lcg(s); thr [base + node] = (int)(s % 256);
        }
        for (int node = LEAF_START; node < NODES; node++) {
            s = lcg(s); leafval[base + node] = (int)(s % 10);
        }
    }
    for (int i = 0; i < N * F; i++) { s = lcg(s); sample[i] = (int)(s % 256); }

    long long _t0 = now_ns();
    pthread_t th[256];
    Band bands[256];
    for (int w = 0; w < cores; w++) {
        bands[w].lo = (int)((long)w * N / cores);
        bands[w].hi = (int)((long)(w + 1) * N / cores);
        pthread_create(&th[w], NULL, worker, &bands[w]);
    }
    for (int w = 0; w < cores; w++) pthread_join(th[w], NULL);
    fprintf(stderr, "COMPUTE_NS %lld\n", now_ns() - _t0);

    long h = 0, total = 0;
    for (int i = 0; i < N; i++) {
        h     = (h     * 31 + accArr[i] + 1) % P;
        total = (total + accArr[i])           % P;
    }

    printf("%ld\n", h);
    printf("gbdt(%d) = %ld\n", N, total);

    free(feat); free(thr); free(leafval); free(sample); free(accArr);
    return 0;
}

File: languages/elixir/gbdt-par.exs
# gbdt-par: parallel scaling-track variant.
# Invocation: elixir gbdt-par.exs <cores> <n>
# Decomposition: partition the N samples into `cores` contiguous bands.
# Each worker traverses all B trees for its samples (trees are read-only, no contention)
# and returns a list of {h_partial, total_partial} for that band's contribution.
# The checksum is assembled serially in sample order after all workers complete.
# Core-invariant: each sample's acc is independent (reads static tree arrays + its own
# feature row). Band boundaries via floor-division are deterministic for any cores.
# Output is identical to the serial benchmark for cores=1,2,4.
import Bitwise

defmodule GbdtPar do
  @p 1_000_000_007
  @d 8
  @b 200
  @f 8
  @nodes 511
  @leaf_start 255

  defp lcg(s), do: band(s * 1_103_515_245 + 12_345, 0x7FFFFFFF)

  defp aget(ref, i), do: :atomics.get(ref, i + 1)
  defp aput(ref, i, v), do: :atomics.put(ref, i + 1, v)

  defp fill_internal(_feat, _thr, _base, node, state) when node >= @leaf_start, do: state

  defp fill_internal(feat, thr, base, node, state) do
    state = lcg(state)
    aput(feat, base + node, rem(state, @f))
    state = lcg(state)
    aput(thr, base + node, rem(state, 256))
    fill_internal(feat, thr, base, node + 1, state)
  end

  defp fill_leaves(_leafval, _base, node, state) when node >= @nodes, do: state

  defp fill_leaves(leafval, base, node, state) do
    state = lcg(state)
    aput(leafval, base + node, rem(state, 10))
    fill_leaves(leafval, base, node + 1, state)
  end

  defp fill_trees(_feat, _thr, _leafval, b, state) when b >= @b, do: state

  defp fill_trees(feat, thr, leafval, b, state) do
    base = b * @nodes
    state = fill_internal(feat, thr, base, 0, state)
    state = fill_leaves(leafval, base, @leaf_start, state)
    fill_trees(feat, thr, leafval, b + 1, state)
  end

  defp fill_sample(_sample, i, total, state) when i >= total, do: state

  defp fill_sample(sample, i, total, state) do
    state = lcg(state)
    aput(sample, i, rem(state, 256))
    fill_sample(sample, i + 1, total, state)
  end

  defp descend(_feat, _thr, _sample, _tbase, _sbase, node, 0), do: node

  defp descend(feat, thr, sample, tbase, sbase, node, depth) do
    f_idx = aget(feat, tbase + node)
    threshold = aget(thr, tbase + node)

    next_node =
      if aget(sample, sbase + f_idx) <= threshold do
        2 * node + 1
      else
        2 * node + 2
      end

    descend(feat, thr, sample, tbase, sbase, next_node, depth - 1)
  end

  defp accum_trees(_feat, _thr, _leafval, _sample, _sbase, b, acc) when b >= @b, do: acc

  defp accum_trees(feat, thr, leafval, sample, sbase, b, acc) do
    tbase = b * @nodes
    leaf = descend(feat, thr, sample, tbase, sbase, 0, @d)
    accum_trees(feat, thr, leafval, sample, sbase, b + 1, acc + aget(leafval, tbase + leaf))
  end

  # Infer for samples [i_start, i_end).
  # Returns a list of acc values (one per sample) in ascending sample order.
  defp infer_band(_feat, _thr, _leafval, _sample, i, i_end, acc_list)
       when i >= i_end,
       do: Enum.reverse(acc_list)

  defp infer_band(feat, thr, leafval, sample, i, i_end, acc_list) do
    sbase = i * @f
    acc = accum_trees(feat, thr, leafval, sample, sbase, 0, 0)
    infer_band(feat, thr, leafval, sample, i + 1, i_end, [acc | acc_list])
  end

  # Compute checksum from a flat list of acc values (in sample order).
  defp checksum([], h, total), do: {h, total}

  defp checksum([acc | rest], h, total) do
    h2 = rem(h * 31 + acc + 1, @p)
    total2 = rem(total + acc, @p)
    checksum(rest, h2, total2)
  end

  def run(cores, n) do
    feat = :atomics.new(@b * @nodes, signed: true)
    thr = :atomics.new(@b * @nodes, signed: true)
    leafval = :atomics.new(@b * @nodes, signed: true)
    sample = :atomics.new(max(n * @f, 1), signed: true)

    state = fill_trees(feat, thr, leafval, 0, 42)
    _state = fill_sample(sample, 0, n * @f, state)

    t0 = System.monotonic_time(:nanosecond)

    # Spawn one Task per band; each returns its list of acc values.
    tasks =
      Enum.map(0..(cores - 1), fn w ->
        i_start = div(w * n, cores)
        i_end = div((w + 1) * n, cores)

        Task.async(fn ->
          infer_band(feat, thr, leafval, sample, i_start, i_end, [])
        end)
      end)

    # Collect results in band order (deterministic), then concatenate into one list.
    acc_values =
      Enum.flat_map(tasks, fn task ->
        Task.await(task, :infinity)
      end)

    t1 = System.monotonic_time(:nanosecond)
    IO.puts(:stderr, "COMPUTE_NS #{t1 - t0}")

    # Serial checksum pass over all samples in order.
    checksum(acc_values, 0, 0)
  end
end

[cores_s, n_s] =
  case System.argv() do
    [c, n | _] -> [c, n]
    [c] -> [c, "5000"]
    _ -> ["1", "5000"]
  end

cores = String.to_integer(cores_s)
n = String.to_integer(n_s)

{h, total} = GbdtPar.run(cores, n)
IO.puts(h)
IO.puts("gbdt(#{n}) = #{total}")
